### PR TITLE
crypto: Add test cases covering the relevant HMAC-SHA{256,512} key length boundaries

### PR DESCRIPTION
**Pull request by practicalswift**
* Add test cases covering the relevant key length boundaries: 64 bytes +/- 1 byte for HMAC-SHA256 and 128 bytes +/- 1 byte for HMAC-SHA512.
* ~~Avoid creating a one-past-the-end pointer in the case of key length 64 (HMAC-SHA256) and key length 128 (HMAC-SHA512).~~
* ~~Avoid performing a noop memset call (zero length argument) in the case of key length 64 (HMAC-SHA256) and key length 128 (HMAC-SHA512).~~

### CODE REVIEW

**Comment by sipa** [comment]
Is this necessary? There is no problem with creating a one-past-end pointer; only dereferencing them, which doesn't happen here.

**Comment by practicalswift** [comment]
@sipa Both zero length `memset`:s and one-past-end pointers are legal, so from a language legality sense this change is not needed.

However, I've seen multiple static analyzers flag this line of code as a potential mistake. This small change makes it apparent for both static analyzers and humans what the intention of the code is. 

My experience is that code that looks like potential mistakes to static analyzers tend to confuse humans too.  I think there is a good case for helping static analyzers in this case. It also has the added benefit of getting rid of an unnecessary `memset` call and getting three additional test cases that cover the key length boundaries.

In summary I think this change is worth it :-)

**Comment by laanwj** [comment]
Pointers to the end of a range are absolutely valid C/C++. This is commonly used for iteration end-exclusive ranges. So is a zero-length memset.

I tend towards NACK adding unnecessary checks here. I think minor harmless changes to fix static analyzer false-positives are sometimes okay, but at some point it goes too far, we don't write our code to appease static analyzers. This kind of ruffling the code can introduce bugs.

File: src/crypto/hmac_sha256.cpp [hunk 1]
```diff
@@ -11,7 +11,9 @@ CHMAC_SHA256::CHMAC_SHA256(const unsigned char* key, size_t keylen)
     unsigned char rkey[64];
     if (keylen <= 64) {
         memcpy(rkey, key, keylen);
-        memset(rkey + keylen, 0, 64 - keylen);
+        if (keylen != 64) {
+            memset(rkey + keylen, 0, 64 - keylen);
+        }
```
Comment (practicalswift) [inline src/crypto/hmac_sha256.cpp]: Removing this change as requested.

File: src/crypto/hmac_sha512.cpp [hunk 2]
```diff
@@ -11,7 +11,9 @@ CHMAC_SHA512::CHMAC_SHA512(const unsigned char* key, size_t keylen)
     unsigned char rkey[128];
     if (keylen <= 128) {
         memcpy(rkey, key, keylen);
-        memset(rkey + keylen, 0, 128 - keylen);
+        if (keylen != 128) {
+            memset(rkey + keylen, 0, 128 - keylen);
+        }
```
Comment (practicalswift) [inline src/crypto/hmac_sha512.cpp]: Removing this change as requested.

**Comment by practicalswift** [comment]
@sipa @laanwj PR updated - now keeping only the new test cases covering the key length boundaries :-)

**Comment by laanwj** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/11516/commits/a3f56578ab79931f3a35a44b6ff43d0da6308602

**Comment by practicalswift** [comment]
Ready for merge? :-)

**Comment by TheBlueMatt** [comment]
Post-merge was-here-looks-fine.
